python/ray/train/tests/test_backend.py

```diff
@@ -106,6 +123,71 @@ def on_shutdown(self, worker_group: WorkerGroup, backend_config: TestConfig):
         pass
 
 
+class MockWorkerGroup(WorkerGroup):
+    # Mock construct metadata
+    def add_workers(self, num_workers: int):
+        """Hard code the WorkerMetadata so that ther will be two nodes, on which
+        even and odd indexed workers are run respectively. For example if there are 4
+        workers, then worker 0 and 2 will be on node 0 and the rest on node 1.
+        """
+        new_actors = []
+        new_actor_metadata = []
+        for i in range(num_workers):
+            actor = self._remote_cls.options(
+                placement_group=self._placement_group
+            ).remote(*self._actor_cls_args, **self._actor_cls_kwargs)
+            new_actors.append(actor)
+            new_actor_metadata.append(
+                WorkerMetadata(
+                    node_id=str(i % 2),
+                    node_ip=str(i % 2),
+                    hostname=str(i % 2),
+                    gpu_ids=[str(i % 2)],
+                )
+            )
+
+        for i in range(len(new_actors)):
+            self.workers.append(
+                Worker(actor=new_actors[i], metadata=new_actor_metadata[i])
+            )
+
+
+class MockBackendExecuter(BackendExecutor):
```

Comment:
We also don't need to create a MockBackendExecutor ourselves- we can leverage the `patch` functionality from `unittest.mock` module for this! https://docs.python.org/3/library/unittest.mock.html.

So if we implement a `mock_construct_metadata`, then we can just do `with path(ray.train._internal.worker_group.construct_data, mock_construct_metadata): `

Reply:
There's some examples of this in `test_backend.py` that you can take a look at